Feature/2534 update apple platform versions

paulrutter opened the pull request:
#2534 Update Apple platform versions
- Added 'OSX_dynamic' and 'OSX_10_16' for Mac OS
- Added 'ipadOS_15_0' / 'ipadOS_15_2'
- Added 'iOS_A_15_0' / 'iOS_A_15_2'
- Added platform to the browsers that it runs on, also including the "generic" platforms to have better coverage for future versions (when the useragent remains the same)

@asgrim @mimmi20 
I did an attempt to add the recent Apple platform versions to browscap.
Also added a "dynamic" OSX item in platforms.json

I still don't exactly understand why those "dynamic" platforms are not added to all browsers, since that would help recognizing future versions which are not specified exactly in "platforms.json". I did that now for the Apple related platforms, but not sure if that's right. Can you help me on this logic? Maybe i'm misunderstanding and those generic ones shouldn't be added?

ocluf committed:
Merge pull request #1 from browscap/6.0.x

updating from browsecap

ocluf committed:
Revert "updating from browsecap"

ocluf committed:
Merge pull request #2 from blueconic/revert-1-6.0.x

Revert "updating from browsecap"

ocluf committed:
Revert "Revert "updating from browsecap""

ocluf committed:
Merge pull request #3 from blueconic/revert-2-revert-1-6.0.x

Revert "Revert "updating from browsecap""

paulrutter committed:
Merge branch 'browscap:6.0.x' into 6.0.x

paulrutter committed:
Merge branch 'browscap:6.0.x' into 6.0.x

paulrutter committed:
Merge branch 'browscap:6.0.x' into 6.0.x

paulrutter committed:
#2534 Update Apple platform versions
- Added 'OSX_dynamic' and 'OSX_10_16' for Mac OS
- Added 'ipadOS_15_0' / 'ipadOS_15_2'
- Added 'iOS_A_15_0' / 'iOS_A_15_2'
- Added platform to the browsers that it runs on, also including the "generic" platforms to have better coverage for future versions (when the useragent remains the same)

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit tests

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit tests

paulrutter committed:
#2534 Update Apple platform versions
- Fix firefox-53-on

paulrutter commented:
Also, can you help determine which unit test if failing? 
I cannot see that in the logs: https://github.com/browscap/browscap/runs/4187437561?check_suite_focus=true

Could it be the "dynamic" platforms that i added to the browser files?

mimmi20 commented on resources/user-agents/browsers/chrome/chrome-ios-73-on.json:
```diff
@@ -24,11 +24,11 @@
           },
           "platforms": [
-            "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
+            "iOS_A_15_2", "iOS_A_15_1", "iOS_A_15_0", "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
```
Please add the new versions (15.x) on a new line

mimmi20 commented on resources/user-agents/browsers/chrome/chrome-ios-73-on.json:
```diff
@@ -37,11 +37,11 @@
           },
           "platforms": [
-            "ipadOS_14_2", "ipadOS_14_1", "ipadOS_14_0",
+            "ipadOS_15_2", "ipadOS_15_1", "ipadOS_15_0", "ipadOS_14_2", "ipadOS_14_1", "ipadOS_14_0",
```
Please add the new versions (15.x) on a new line

mimmi20 commented on resources/user-agents/browsers/mobile-safari/mobile-safari-generic.json:
```diff
@@ -22,7 +22,7 @@
           },
           "platforms": [
-            "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
+            "iOS_A_15_2", "iOS_A_15_1", "iOS_A_15_0", "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
```
Please add the new versions (15.x) on a new line

paulrutter committed:
#2534 Update Apple platform versions
- Code review comments

paulrutter commented:
I fixed the comments, but i think that won't help solve the unit test issue.
Can you take a look at my question here? https://github.com/browscap/browscap/pull/2535#issuecomment-966912527 and https://github.com/browscap/browscap/pull/2535#issue-779165451

paulrutter commented:
Thanks for approving @mimmi20, but since the unit tests still fail i don't think we should merge.
I need more info on the "dynamic" platforms, as mentioned earlier.

mimmi20 commented:
"dynamic" platforms only work, if the platform version matches the browser version.

mimmi20 commented:
The Build is failing since May. Idont think your changes caused this.

paulrutter committed:
#2534 Update Apple platform versions
- Attempt to fix unit test by reducing expectation to 1

paulrutter commented:
> The Build is failing since May. Idont think your changes caused this.

I will try to fix the unit tests first, so we can actually see what my changes did to existing functionality.

paulrutter committed:
#2534 Update Apple platform versions
- Attempt to fix unit test by reducing expectation to 1

paulrutter committed:
#2534 Update Apple platform versions
- Attempt to fix unit test by reducing expectation to 1

paulrutter committed:
#2534 Update Apple platform versions
- Attempt to fix unit test by reverting to any

paulrutter committed:
#2534 Update Apple platform versions
- Update status badge for CI integration

paulrutter committed:
#2534 Update Apple platform versions
- Revert dynamic platforms, as these don't work

paulrutter committed:
#2534 Update Apple platform versions
- Revert dynamic

paulrutter committed:
#2534 Update Apple platform versions
- Add safari 15 as that's included with Mac OS X 15

paulrutter committed:
#2534 Update Apple platform versions
- Add mobile safari minor versions that exist

paulrutter committed:
#2534 Update Apple platform versions
- Add mobile safari minor versions that exist for 13

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit test now version is therer

paulrutter committed:
#2534 Update Apple platform versions
- Add Mac OS X 11 (Big sur) and 12 (Monterey)

paulrutter committed:
#2534 Update Apple platform versions
- Correct device type

paulrutter committed:
#2534 Update Apple platform versions
- correct platform

paulrutter committed:
#2534 Update Apple platform versions
- correct platform

paulrutter committed:
#2534 Update Apple platform versions
- correct platform

paulrutter committed:
#2534 Update Apple platform versions
- Add missing iOS platform versions

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit test

paulrutter committed:
#2534 Update Apple platform versions
- Add big sur and monterey

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit test

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit test

paulrutter committed:
#2534 Update Apple platform versions
- Updated unit tests

paulrutter committed:
#2534 Update Apple platform versions
- Allow opera on newer mac os x

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit tests
- Edge using different version numbers from 75

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit tests

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit tests

paulrutter committed:
#2534 Update Apple platform versions
- Fix unit tests

mimmi20 commented on resources/user-agents/browsers/chrome/chrome-generic.json:
```diff
@@ -48,7 +48,9 @@
           "platforms": [
-            "OSX_10_15", "OSX_10_14", "OSX_10_13", "OSX_10_12",
+            "OSX_12_0",
+            "OSX_11_0", "OSX_11_1",
```
Please change the ordering. Please write `"OSX_11_1", "OSX_11_0"` (like you did in the Chrome file). Please change it in other files too.

AnuarT79 replied:
> Please change the ordering. Please write `"OSX_11_1", "OSX_11_0"` (like you did in the Chrome

mimmi20 commented on resources/user-agents/browsers/mobile-safari/mobile-safari-15-0.json:
```diff
@@ -0,0 +1,66 @@
+{
+  "division": "Mobile Safari #MAJORVER#.#MINORVER#",
+  "versions": [14],
```
Please change the version to 15

mimmi20 commented on resources/user-agents/browsers/mobile-safari/mobile-safari-14-0.json:
```diff
@@ -1,6 +1,6 @@
 {
   "division": "Mobile Safari #MAJORVER#.#MINORVER#",
-  "versions": [14],
+  "versions": ["14.0", "14.1", "14.2", "14.3", "14.4", "14.5", "14.6", 14],
```
Please change the order of versions. Please write the highest first. The version "14.0" is not needed, as it the same as version "14". Also change the file name to `mobile-safari-14-x.json`.

mimmi20 commented on resources/user-agents/browsers/mobile-safari/mobile-safari-13-0.json:
```diff
@@ -1,6 +1,6 @@
 {
   "division": "Mobile Safari #MAJORVER#.#MINORVER#",
-  "versions": [13],
+  "versions": ["13.1", 13],
```
Please change the filename to `mobile-safari-13-x.json`.

mimmi20 commented on resources/user-agents/browsers/safari/safari-15-0.json:
```diff
@@ -0,0 +1,33 @@
+{
+    "division": "Safari #MAJORVER#.#MINORVER#",
+    "versions": ["15.0", 15],
```
These two versions are the same, as the Zero will be added automatically, if a `#MINORVER#` placeholder is there.

mimmi20 commented on tests/BrowscapTest/Generator/Helper/BuildHelperTest.php:
```diff
@@ -784,13 +784,13 @@ public function testDuplicateUseragents(): void
             ->method('renderAllDivisionsHeader')
             ->willReturnSelf();
-        $writerCollection->expects(static::exactly(2))
+            $writerCollection->expects(static::any())
```
Please keep the indent

paulrutter committed:
#2534 Update Apple platform versions
- Code review changes

paulrutter commented:
@mimmi20 I managed to get the unit tests running again, by reverting a change in https://github.com/browscap/browscap/pull/2535/files#diff-cebc14b7503943d2278cbaf4fe86c60f57707b619511f54041399054709dd28d.

Now i had to fix few other unit tests that were added already, but didn't succeed.
I've tried, to my best knowledge, to add the new Apple platforms:

* Mac OS Big sur (11) and Monterey (12), 10.x range is no longer used although 10.16 was used at some point
* iOS 14.0 / 14.8 and 15.0 / 15.2 (which is in beta)
* iPad OS 15.0 / 15.2, didn't add 14.3 / 14.8 as i wasn't sure it was used

I also added Safari 15 specifically, as that is shipped with MacOS Monterey. 
In the readme, i updated the badge for the CI, which was no longer ok.

@mimmi20 i processed your latest comments. Can you re-review and merge if you think it's ok now?
The code coverage report still seems a bit off, but not sure how to resolve that.

paulrutter committed:
#2534 Update Apple platform versions
- Update unit test

paulrutter committed:
#2534 Update Apple platform versions
- Updated README.md

paulrutter commented:
@mimmi20 can you merge and release a new browscap version if you're ok with the changes?

mimmi20 commented:
@paulrutter I may be able to merge, but not to release.
@asgrim Could you merge and release?

asgrim commented:
Yep, thank you @paulrutter - appreciate the effort and fixing up those tests :metal:

paulrutter commented:
@asgrim do we need to trigger something for http://browscap.org/ as well?

asgrim commented:
Yes, waiting for Dependabot

@mimmi20 FYI the release procedure is at https://github.com/browscap/browscap/wiki/Public-release-procedure :) I believe you should be able to do most of it now

asgrim commented:
@paulrutter looks like it is up now :+1:

paulrutter commented:
Yes, great!

willyaranda commented:
`Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Safari/605.1.15`

is incorrectly detected, as does not match. Changing `Version/15.1` for `Version/15.0` does match and get the major version properly

paulrutter commented:
Hmm i wasn't aware of safari 15.1 was already released. We might add that as well.

paulrutter commented:
Created https://github.com/browscap/browscap/pull/2536 @willyaranda

AnuarT79 commented:
Obvious
